airbyte-integrations/connectors/source-linkedin-ads/source_linkedin_ads/source.py

```diff
@@ -0,0 +1,352 @@
+        To get the next page, set start value to N, while the count value stays the same.
+        We have reached the end of the dataset when the response contains fewer elements than the `count` parameter request.
+        https://docs.microsoft.com/en-us/linkedin/shared/api-guide/concepts/pagination?context=linkedin/marketing/context
+        """
+        if len(response.json().get("elements")) < self.limit:
+            return None
+        return {"start": response.json().get("paging").get("start") + self.limit}
+
+    def request_params(
+        self, stream_state: Mapping[str, Any], next_page_token: Mapping[str, Any] = None, **kwargs
+    ) -> MutableMapping[str, Any]:
+        params = {"count": self.limit, "q": "search"}
+        if next_page_token:
+            params.update(**next_page_token)
+        return params
+
+    def parse_response(self, response: requests.Response, **kwargs) -> Iterable[Mapping]:
+        """
+        We need to get out the nested complex data structures for further normalisation, so the transform_data method is applied.
+        """
+        yield from transform_data(response.json().get("elements"))
+
+
+class Accounts(LinkedinAdsStream):
+    """
+    Get Accounts data. More info about LinkedIn Ads / Accounts:
+    https://docs.microsoft.com/en-us/linkedin/marketing/integrations/ads/account-structure/create-and-manage-accounts?tabs=http
+    """
+
+    endpoint = "adAccountsV2"
+
+    def request_headers(self, stream_state: Mapping[str, Any], **kwargs) -> Mapping[str, Any]:
+        """
+        If account_ids are specified as user's input from configuration,
+        we must use MODIFIED header: {'X-RestLi-Protocol-Version': '2.0.0'}
+        """
+        return {"X-RestLi-Protocol-Version": "2.0.0"} if self.accounts else {}
+
+    def request_params(self, stream_state: Mapping[str, Any], **kwargs) -> MutableMapping[str, Any]:
+        """
+        Override request_params() to have the ability to accept the specific account_ids from user's configuration.
+        If we have list of account_ids, we need to make sure that the request_params are encoded correctly,
+        We will get HTTP Error 500, if we use standard requests.urlencode methods to parse parameters,
+        so the urlencode(..., safe=":(),") is used instead, to keep the values as they are.
+        """
+        params = super().request_params(stream_state=stream_state, **kwargs)
+        if self.accounts:
+            account_list = ",".join(map(str, self.accounts))
+            params["search"] = f"(id:(values:List({account_list})))"
+            return urlencode(params, safe=":(),")
+        return params
+
+
+class IncrementalLinkedinAdsStream(LinkedinAdsStream):
+
+    cursor_field = "lastModified"
+
+    @property
+    def primary_slice_key(self) -> str:
+        """
+        Define the main slice_key from `slice_key_value_map`. Always the first element.
```

Comment:
It's purpose to simplify the process of getting the slice for all dependent streams that are using `StreamMixin` class.
For each class, we now define the mapping (if required) such as: 
```
slice_key_value_map = {"account_id": "id"}
```
When we form the slice for the dependent stream, like Campaigns (depends on Accounts), we pass the `id` of the account to each request, to find all the campaigns associated with the specific `account_id`, instead of overriding the read_records() method for each stream, we simply defining the `slice_key_value_map` and keep the read_records in the top class.